contributors/design-proposals/storage/dynamic_volume_limit.md

```diff
@@ -0,0 +1,102 @@
+  the feature `DynamicVolumeThreshold`.
+
+### API Changes
+
+There is no API change needed for this feature. However existing `node.Status.Capacity` will
+be extended to cover volume limits available on the node too.
+
+The key name that will store volume will be start with prefix `storage-`. So for example:
+
+* `storage-aws-ebs`
+* `storage-gce-pd`
+* `storage-csi-<driver_name>`
+
+`IsScalarResourceName` check will be extended to cover storage limits:
+
+```go
+func IsStorageResourceName(name v1.ResourceName) bool {
+    return strings.HasPrefix(string(name), v1.ResourceStoragePrefix)
+}
+
+// Extended and Hugepages resources
+func IsScalarResourceName(name v1.ResourceName) bool {
+    return IsExtendedResourceName(name) || IsHugePageResourceName(name) ||
+        IsPrefixedNativeResource(name) || IsStorageResourceName(name)
+}
+```
+
+#### Alternative to using "storage-" prefix
+We also considered using currently defined `GetPluginName` interface(of Volume Plugins) for using as key in the `node.Status.Capacity`. Ultimately
+we decided against using it, because most in-tree plugins start with `kubernetes.io/` and we needed a uniform way to identify storage
+related capacity data in `node.Status`.
+
+### Changes to CSI
+
+Corresponding change is being made to CSI so as CSI drivers can output
+volume limits as part of Node capabilities - https://github.com/container-storage-interface/spec/pull/221
+
+csi node-registrar will be updated to set the `node.Status` property after fetching
+volume limit from CSI driver.
+
+### Changes to scheduler
+
+Scheduler will retrieve available attachable limit on a node from `node.Status.Capacity` and store it in `nodeInfo` cache. Volume
+limits will be treated like any other scalar resource.
+
+Existing `NewMaxPDVolumeCountPredicate` will be extended to calculate volumes in-use on a node by counting all scheduled pods on the node
+and then comparing it with limit available from `nodeInfo`.
+
+To be backward compatible - the scheduler will fallback to older logic, if no limit is set in `node.Status.Capacity` for AWS, GCE and Azure volume
+types. This is to make sure that - a updated scheduler can still work with older nodes.
+
+### Setting of limit for existing in-tree volume plugins
+
+The volume limit for existing volume plugins will be set by querying the volume plugin. Following function
+will be added to volume plugin interface:
+
+```go
+type VolumePlugin interface {
+    GetVolumeLimits() (map[string]int64, error)
+}
+```
+
+When querying the plugin - plugin will use `ProviderName` function of CloudProvider to check
+if plugin is usable on the node. For example - querying for `GetVolumeLimits` from `aws-ebs` plugin with `gce` cloudprovider
+will result in error.
+
+Kubelet will query the volume plugins inside `kubelet.initialNode` function and populate `node.Status` with returned values.
```

Comment:
I am assuming that node.Status will be populated at the time of node registration before node is ready and before pods can be scheduled on it. I am assuming the limit wont change at run time, right?

Reply:
Yes limit will not change at runtime and should be invariant through out lifetime of node.